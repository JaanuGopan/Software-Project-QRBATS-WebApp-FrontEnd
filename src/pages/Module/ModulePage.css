@media screen and (max-width: 1800px) {
  .module-mainHead {
    font-size: x-large;
    margin-top: 5px;
    padding-left: 35px;
  }

  .module-buttonIcon {
    height: 1.5rem;
    width: 1.5rem;
    size: 1.5rem;
  }

  .module-Dash {
    margin-top: 20px;
  }

  .module-SearchEvent {
    margin-right: 35px;
  }
}

@media screen and (max-width: 768px) {
  .module-mainHead {
    font-size: large;
    margin-top: 10px;
    padding-left: 35px;
  }

  .module-buttonIcon {
    height: 1rem;
    width: 1rem;
    size: 1rem;
  }

  .module-Dash {
    margin-top: 10px;
  }

  .module-SearchEvent {
    margin-right: 35px;
  }
}

@media screen and (max-width: 426px) {
  .module-buttonIcon {
    height: 0.5rem;
    width: 0.5rem;
    size: 0.5rem;
  }

  .module-Dash {
    margin-top: 5px;
  }

  .module-SearchEvent {
    margin-right: 0px;
  }

  .module-mainHead {
    padding-left: 5px;
  }
}

.module-create-or-update-modal-container {
  position: fixed;
  z-index: 4;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
}

.module-Dash {
  flex-grow: 1;
  height: 85vh;
  width: 100%;
  z-index: 1;
}

.module-mainHead {
  color: #0d295e;
  font-weight: bold;
  margin-bottom: 0px;
}

.module-ModuleList {
  border-top: 5px solid #0365a1;
  box-shadow: 0.5px 2px 0.5px 0.5px rgba(25, 65, 85, 0.5);
  border-radius: 10px;
  background-color: #fafafa;
  max-height: calc(100vh - 220px);
}

.module-List-Buttons {
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: space-around;
  padding: 1rem;
}

.module-SearchEvent {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-between;
  padding-right: 10px;
}

.module-create-main-container {
  border: 4px solid transparent;
  border-top: 6px solid #0077c0;
  border-radius: 10px;
  background-color: white;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);
  z-index: 2;
}

.module-update-main-container {
  position: fixed;
  width: 50%;
  border: 4px solid transparent;
  border-top: 6px solid #0077c0;
  border-radius: 10px;
  padding: 20px;
  background-color: white;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);
  min-width: 250px;
  z-index: 2;
}

.module-head1 {
  text-align: center;
  color: black;
  font-weight: bold;
  font-size: x-large;
  padding: 1%;
  margin: 0%;
}

.module-login-container {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: space-between;
  align-items: center;
  margin-top: 20px;
  position: relative;
}

.module-image-container {
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.module-logo {
  width: 200px;
  height: auto;
  border-radius: 10px;
}

.module-select-input {
  border-radius: 10px;
  word-wrap: normal;
  padding: 5px;
  width: 100%;
  border-bottom: 3px solid #0063a0;
}

.closeCreateModuleWindow {
  color: red;
  cursor: pointer;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  align-items: center;
  justify-content: center;
  padding: 5px;
  border-radius: 50px;
  margin-left: 95%;
  width: 30px;
  height: 30px;
}

.closeCreateModuleWindow:hover {
  color: white;
  background-color: red;
}

.module-creation-input {
  display: flex;
  flex-direction: row;
  width: 100%;
  align-items: center;
  justify-content: space-between;
  align-self: center;
  justify-self: center;
  margin: auto;
  margin-top: 1%;
  margin-bottom: 1%;
  padding-left: 5%;
  padding-right: 5%;
}

.module-creation-input label {
  width: 30%;
}

.module-creation-input-field {
  width: 60%;
}

.create-module-create-button {
  display: flex;
  flex-direction: row;
  width: 100%;
  align-self: center;
  justify-self: center;
  margin-top: 20px;
  align-self: center;
  justify-content: center;
}

.create-module-create-button button {
  width: 40%;
  align-self: center;
  justify-self: center;
}

.module-create-title-close-button {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  width: 100%;
}

.module-create-title {
  align-self: center;
  justify-self: center;
  margin-left: auto;
}

.module-create-close-button {
  align-self: center;
  justify-self: flex-end;
  margin-left: auto;
  height: 30px;
  width: 30px;
  align-items: center;
  justify-content: center;
  margin-bottom: 30px;
}

.module-create-close-button #close-icon {
  align-self: center;
  justify-self: center;
  color: red;
  height: 90%;
  width: 90%;
}

.module-create-close-button #close-icon:hover {
  align-self: center;
  justify-self: center;
  color: white;
  height: 90%;
  width: 90%;
  background-color: red;
  border-radius: 100%;
}

.module-delete-window-container {
  position: fixed;
  z-index: 4;
  left: 0;
  top: 0;
  overflow-y: scroll;
  width: 100%;
  height: 100vh;
  padding-top: 5vh;
  padding-bottom: 5vh;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.5);
}
