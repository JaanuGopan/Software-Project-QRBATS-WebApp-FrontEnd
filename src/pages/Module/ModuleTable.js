import React, { useState, useEffect } from "react";
import { useSelector } from "react-redux";
import { selectUser } from "../../redux/features/userSlice";
import ModuleService from "../../api/services/ModuleService";
import { CiViewList } from "react-icons/ci";
import "../AdminDashboard/AdminDashboard.css";
import { BiEdit } from "react-icons/bi";

const ModuleTable = ({
  handleOpenCreateModuleWindow,
  handleOpenUpdateModuleWindow,
  searchModule,
  onModuleClick,
  modulesList,
  handleDeleteModule,
}) => {
  const user = useSelector(selectUser);
  const { userId } = user || {};
  const [moduleList, setModuleList] = useState([]);
  const [selectedModule, setSelectedModule] = useState(null);

  const departmentList = ["DEIE", "DCOM", "DMME", "DCEE", "DMENA"];

  useEffect(() => {
    setModuleList(modulesList);
  }, [modulesList]);

  const handleModuleClick = (module) => {
    setSelectedModule(module);
    onModuleClick(module);
  };

  const handleReloadModuleList = async (userId) => {
    ModuleService.getModulesByUserId(userId).then((modules) => {
      setModuleList(modules);
    });
  };

  return (
    <div className="tableDesign">
      <table className="event-report-tableArrangement">
        <thead>
          <tr>
            <th>No</th>
            <th className="expand">Module Name</th>
            <th>Module Code</th>
            <th>Semester</th>
            <th>Department</th>
            <th>Edit</th>
          </tr>
        </thead>
        <tbody>
          {moduleList
            .filter(
              (module) =>
                module.moduleName
                  .toLowerCase()
                  .includes(searchModule.toLowerCase()) ||
                module.moduleCode
                  .toLowerCase()
                  .includes(searchModule.toLowerCase())
            )
            .map((module, index) => (
              <tr
                key={index}
                onClick={() => handleModuleClick(module)}
                className={
                  selectedModule === module ? "selected-row" : "event-row"
                }
              >
                <td>{index + 1}</td>
                <td>{module.moduleName}</td>
                <td>{module.moduleCode}</td>
                <td>{module.semester}</td>
                <td>{departmentList[module.departmentId - 1]}</td>
                <td>
                  <button
                    onClick={handleOpenUpdateModuleWindow}
                    className="ViewButton"
                  >
                    <BiEdit className="EditIcon" />
                    <p className="ViewButtonLabel">Edit</p>
                  </button>
                </td>
              </tr>
            ))}
        </tbody>
      </table>
    </div>
  );
};

export default ModuleTable;
