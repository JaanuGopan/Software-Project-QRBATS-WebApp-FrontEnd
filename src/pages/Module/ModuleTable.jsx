import React, { useState, useEffect } from 'react';
import ModuleService from '../../api/services/ModuleService';
import '../AdminDashboard/AdminDashboard.css';
import { BiEdit } from 'react-icons/bi';
import Department from '../../utils/Department';

const ModuleTable = ({
  handleOpenCreateModuleWindow,
  handleOpenUpdateModuleWindow,
  searchModule,
  onModuleClick,
  modulesList,
  handleDeleteModule,
}) => {
  const [moduleList, setModuleList] = useState([]);
  const [selectedModule, setSelectedModule] = useState(null);

  const departmentList = Department.studentDepartmentList;

  useEffect(() => {
    setModuleList(modulesList);
  }, [modulesList]);

  const handleModuleClick = (module) => {
    setSelectedModule(module);
    onModuleClick(module);
  };

  const handleReloadModuleList = async (userId) => {
    ModuleService.getModulesByUserId(userId).then((modules) => {
      setModuleList(modules);
    });
  };

  return (
    <div className="tableDesign">
      <table className="event-report-tableArrangement">
        <thead>
          <tr>
            <th>No</th>
            <th className="expand">Module Name</th>
            <th>Module Code</th>
            <th>Semester</th>
            <th>Department</th>
            <th>Edit</th>
          </tr>
        </thead>
        {moduleList.filter(
          (module) =>
            module.moduleName.toLowerCase().includes(searchModule.toLowerCase()) ||
            module.moduleCode.toLowerCase().includes(searchModule.toLowerCase())
        ).length > 0 ? (
          <tbody>
            {moduleList
              .filter(
                (module) =>
                  module.moduleName.toLowerCase().includes(searchModule.toLowerCase()) ||
                  module.moduleCode.toLowerCase().includes(searchModule.toLowerCase())
              )
              .map((module, index) => (
                <tr
                  key={index}
                  onClick={() => handleModuleClick(module)}
                  className={selectedModule === module ? 'selected-row' : 'event-row'}
                >
                  <td>{index + 1}</td>
                  <td>{module.moduleName}</td>
                  <td>{module.moduleCode}</td>
                  <td>{module.semester}</td>
                  <td>{departmentList[module.departmentId - 1]}</td>
                  <td>
                    <button onClick={handleOpenUpdateModuleWindow} className="ViewButton">
                      <BiEdit className="EditIcon" />
                      <p className="ViewButtonLabel">Edit</p>
                    </button>
                  </td>
                </tr>
              ))}
          </tbody>
        ) : (
          <tbody className="nodata">No Data Available</tbody>
        )}
      </table>
    </div>
  );
};

export default ModuleTable;
