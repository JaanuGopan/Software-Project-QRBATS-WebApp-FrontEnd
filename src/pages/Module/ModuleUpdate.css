.module-update-container {
    position: fixed;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    z-index: 9999;
    /* Set a high z-index value to bring it to the front */
    background-color: white;
    /* Add background color or styles as needed */
    width: 400px;
    /* Set width according to your design */
    padding: 20px;
    /* Add padding as needed */
    border: 1px solid #ccc;
    /* Add border styles as needed */
    border-radius: 5px;
    /* Add border radius as needed */
}