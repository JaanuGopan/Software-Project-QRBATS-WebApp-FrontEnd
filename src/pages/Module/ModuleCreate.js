import React, { useState } from "react";
import "./ModulePage.css";
import Designer from "../../assets/Images/Designer.jpeg";
import InputField from "../../components/textfields/InputBox/InputField";
import { IoMdCloseCircleOutline } from "react-icons/io";
import CreateUserService from "../../api/services/CreateUserService";
import InputList from "../../components/textfields/InputList/InputList";

const ModuleCreate = ({ handleCloseModuleCreateWindow }) => {
  const [moduleName, setModuleName] = useState("");
  const [moduleCode, setModuleCode] = useState("");
  const [semester, setSemester] = useState("");
  const [moduleEnrolmentKey, setModuleEnrolmentKey] = useState("");
  const [departmentId, setDepartmentId] = useState("");
  const [lectureId, setLectureId] = useState("");

  const departmentList = ["DEIE", "DCOM", "DMME", "DCEE", "DMENA"];
  const semesterList = ["1", "2", "3", "4", "5", "6", "7", "8"];

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      const response = CreateUserService.saveUser();

      console.error("Login success...");
      handleCloseModuleCreateWindow();
    } catch (error) {
      console.error("Login failed", error);
    }
  };

  return (
    <div className="module-signup-main-container">
      <div
        className="closeCreateEventWindow"
        onClick={handleCloseModuleCreateWindow}
      >
        <IoMdCloseCircleOutline />
      </div>
      <p className="module-head1">Create module</p>
      <div className="module-login-container">
        <div className="module-image-container">
          <img src={Designer} className="module-logo" alt="Logo" />
        </div>
        <div className="form-container">
          <form onSubmit={handleSubmit}>
            {/* First Name Input */}
            <InputField
              placeholder="Enter The Module Name"
              value={moduleName}
              onChange={(e) => setModuleName(e.target.value)}
              inputType="text"
            />

            {/* Last Name Input */}
            <InputField
              placeholder="Enter the module code"
              value={moduleCode}
              onChange={(e) => setModuleCode(e.target.value)}
              inputType="text"
            />

            <div className="choice-input mb-3">
              <InputList
                placeholder="Select semester"
                value={semester}
                onChange={(e) => setSemester(e.target.value)}
                list={semesterList}
                inputType="text"
                initialValue="Select semester"
                className="student-select-input"
              />
            </div>

            <div className="choice-input mb-3">
              <InputList
                placeholder="Select department"
                value={departmentId}
                onChange={(e) => setDepartmentId(e.target.value)}
                list={departmentList}
                inputType="text"
                initialValue="Select department"
                className="student-select-input"
              />
            </div>

            {/* Password Input */}
            <InputField
              placeholder="Enter Module Enrolment Key"
              value={moduleEnrolmentKey}
              onChange={(e) => setModuleEnrolmentKey(e.target.value)}
              inputType="password"
            />

            <button type="submit" className="btn btn-success w-100">
              Create module
            </button>
          </form>
        </div>
      </div>
    </div>
  );
};

export default ModuleCreate;
