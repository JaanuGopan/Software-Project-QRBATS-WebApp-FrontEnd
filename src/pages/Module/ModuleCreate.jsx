import React, { useState } from 'react';
import './ModulePage.css';
import Designer from '../../assets/Images/Designer.jpeg';
import InputField from '../../components/textfields/InputBox/InputField';
import { IoMdCloseCircleOutline } from 'react-icons/io';
import { IoIosClose } from 'react-icons/io';
import InputList from '../../components/textfields/InputList/InputList';
import ModuleService from '../../api/services/ModuleService';
import { selectUser } from '../../redux/features/userSlice';
import { useSelector } from 'react-redux';
import { ToastContainer, toast } from 'react-toastify';
import Select from 'react-select';
import { CircularProgress } from '@mui/material';

const ModuleCreate = ({
  handleCloseModuleCreateWindow,
  handleReloadModuleList,
  handleModuleCreate,
}) => {
  const [moduleName, setModuleName] = useState('');
  const [moduleCode, setModuleCode] = useState('');
  const [semester, setSemester] = useState('');
  const [moduleEnrolmentKey, setModuleEnrolmentKey] = useState('');
  const [departmentId, setDepartmentId] = useState('');

  const { userId } = useSelector(selectUser) || {};

  const departmentList = ['DEIE', 'DCOM', 'DMME', 'DCEE', 'DMENA', 'DIS'];
  const semesterList = ['1', '2', '3', '4', '5', '6', '7', '8'];

  const [processing, setProcessing] = useState(false);

  const handleCreateModule = async (e) => {
    e.preventDefault();
    const moduleData = {
      moduleCode: moduleCode,
      moduleName: moduleName,
      semester: semester,
      moduleEnrolmentKey: moduleEnrolmentKey,
      departmentId: departmentList.indexOf(departmentId) + 1,
      userId,
    };
    try {
      setProcessing(true);
      await handleModuleCreate(moduleData);
    } finally {
      setProcessing(false);
    }
  };

  const handleClearData = () => {
    setModuleCode('');
    setModuleName('');
    setDepartmentId('');
    setModuleEnrolmentKey('');
    setSemester('');
  };

  return (
    <div className="module-create-main-container">
      <div className="module-create-title-close-button">
        <h3 className="module-create-title">Create Module</h3>
        <div className="module-create-close-button" onClick={handleCloseModuleCreateWindow}>
          <IoMdCloseCircleOutline id="close-icon" />
        </div>
      </div>
      <div className="module-login-container">
        <div className="form-container">
          <form onSubmit={handleCreateModule}>
            <div className="module-creation-input">
              <label>Module Name</label>
              <div className="module-creation-input-field">
                <InputField
                  placeholder="Enter The Module Name"
                  value={moduleName}
                  onChange={(e) => setModuleName(e.target.value)}
                  inputType="text"
                />
              </div>
            </div>

            <div className="module-creation-input">
              <label>Module Code</label>
              <div className="module-creation-input-field">
                <InputField
                  placeholder="Enter the module code"
                  value={moduleCode}
                  onChange={(e) => setModuleCode(e.target.value)}
                  inputType="text"
                />
              </div>
            </div>

            <div className="module-creation-input">
              <label>Semester</label>
              <div className="module-creation-input-field">
                <div className="choice-input mb-3">
                  <InputList
                    placeholder="Select semester"
                    value={semester}
                    onChange={(e) => setSemester(e.target.value)}
                    list={semesterList}
                    inputType="text"
                    initialValue="Select semester"
                    className="student-select-input"
                  />
                </div>
              </div>
            </div>

            <div className="module-creation-input">
              <label>Department</label>
              <div className="module-creation-input-field">
                <div className="choice-input mb-3">
                  <InputList
                    placeholder="Select department"
                    value={departmentId}
                    onChange={(e) => setDepartmentId(e.target.value)}
                    list={departmentList}
                    inputType="text"
                    initialValue="Select department"
                    className="student-select-input"
                  />
                </div>
              </div>
            </div>

            <div className="module-creation-input">
              <label>Enrolment Key</label>
              <div className="module-creation-input-field">
                <InputField
                  placeholder="Enter Module Enrolment Key"
                  value={moduleEnrolmentKey}
                  onChange={(e) => setModuleEnrolmentKey(e.target.value)}
                  inputType="text"
                />
              </div>
            </div>
            <div className="create-module-create-button">
              {processing ? (
                <CircularProgress />
              ) : (
                <button type="submit" className="btn btn-success">
                  Create module
                </button>
              )}
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};

export default ModuleCreate;
