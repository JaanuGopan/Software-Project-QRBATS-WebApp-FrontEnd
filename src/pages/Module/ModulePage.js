import React, { useState, useEffect } from "react";
import NormalButton from "../../components/layout/AdminDashboardComponent/NormalButton";
import { MdArrowBack, MdCreateNewFolder } from "react-icons/md";
import { BiSolidPrinter } from "react-icons/bi";
import ModuleTable from "./ModuleTable";
import ModuleService from "../../api/services/ModuleService";
import { useSelector } from "react-redux";
import { selectUser } from "../../redux/features/userSlice";
import ModuleUpdate from "./ModuleUpdate";
import { RiDeleteBin5Fill } from "react-icons/ri";
import "./ModulePage.css";
import ModuleCreate from "./ModuleCreate";

const ModulePage = () => {
  const user = useSelector(selectUser);
  const { userId } = user || {};

  const [showModuleTable, setShowModuleTable] = useState(true);
  const [showUpdateModuleWindow, setShowUpdateModuleWindow] = useState(false);
  const [showModuleCreateWindow, setShowModuleCreateWindow] = useState(false);

  const [selectedModule, setSelectedModule] = useState(null);
  const [moduleList, setModuleList] = useState([]);
  const [searchModule, setSearchModule] = useState("");

  useEffect(() => {
    handleReloadModuleList();
  }, []);

  const handleModuleClick = (module) => {
    setSelectedModule(module);
    console.log("Selected Module:", module);
  };

  const handleReloadModuleList = async () => {
    ModuleService.getModulesByUserId(userId).then((modules) => {
      setModuleList(modules);
      console.log(modules);
    });
  };

  const handleOpenUpdateModuleWindow = () => {
    setShowModuleTable(false);
    setShowUpdateModuleWindow(true);
  };

  return (
    <div>
      <div className="module-Dash">
        <div className="module-SearchEvent">
          <p className="module-mainHead">Modules</p>
          <input
            type="text"
            placeholder="Search..."
            style={{
              width: "150px",
              padding: "3px 40px",
              border: "0.5px solid black",
              borderRadius: "5px",
              textAlign: "center",
            }}
            onChange={(e) => setSearchModule(e.target.value)}
          />
        </div>
        <div className="module-EventList">
          <ModuleTable
            modulesList={moduleList}
            onModuleClick={handleModuleClick}
            searchModule={searchModule}
            handleDeleteModule={() => {}}
            handleOpenCreateModuleWindow={() => {}}
            handleOpenUpdateModuleWindow={() => {
              setShowUpdateModuleWindow(true);
            }}
          />

          <div className="module-List-Buttons">
            <NormalButton
              handleClick={() => {
                setShowModuleCreateWindow(true);
              }}
              title={"Create"}
              titlewithiconicon={<MdCreateNewFolder className="buttonIcon" />}
            />
            <NormalButton
              title={"Delete"}
              handleClick={() => {}}
              titlewithiconicon={<RiDeleteBin5Fill className="buttonIcon" />}
            />
          </div>
        </div>
        {showUpdateModuleWindow && (
          <div
            className="event-Create-Event-Dashboard"
            handleClick={() => {
              setShowModuleCreateWindow(false);
            }}
          >
            <ModuleUpdate
              selectedModule={selectedModule}
              handleCloseUpdateWindow={() => {}}
              handleUpdateModule={() => {}}
              reloadModuleList={handleReloadModuleList}
            />
          </div>
        )}
        {showModuleCreateWindow && (
          <div
            className="event-Create-Event-Dashboard"
            handleClick={() => {
              setShowUpdateModuleWindow(false);
            }}
          >
            <ModuleCreate
              handleCloseModuleCreateWindow={() => {
                setShowModuleCreateWindow(false);
              }}
            />
          </div>
        )}
      </div>
    </div>
  );
};

export default ModulePage;
