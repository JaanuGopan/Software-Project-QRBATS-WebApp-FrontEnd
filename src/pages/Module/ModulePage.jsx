import React, { useState, useEffect, useContext } from 'react';
import NormalButton from '../../components/layout/AdminDashboardComponent/NormalButton';
import { MdCreateNewFolder } from 'react-icons/md';
import ModuleTable from './ModuleTable';
import ModuleService from '../../api/services/ModuleService';
import ModuleUpdate from './ModuleUpdate';
import { RiDeleteBin5Fill } from 'react-icons/ri';
import './ModulePage.css';
import ModuleCreate from './ModuleCreate';
import { ToastContainer, toast } from 'react-toastify';
import WarningPopup from '../../components/warningPopup/WarningPopup';
import { AuthContext } from '../../config/AuthProvider';

const ModulePage = () => {
  const { user } = useContext(AuthContext);
  const { userId } = user || {};

  const [showModuleTable, setShowModuleTable] = useState(true);
  const [showUpdateModuleWindow, setShowUpdateModuleWindow] = useState(false);
  const [showModuleCreateWindow, setShowModuleCreateWindow] = useState(false);
  const [showDeleteModuleWindow, setShowDeleteModuleWindow] = useState(false);

  const [selectedModule, setSelectedModule] = useState(null);
  const [moduleList, setModuleList] = useState([]);
  const [searchModule, setSearchModule] = useState('');

  useEffect(() => {
    handleReloadModuleList();
  }, []);

  const handleModuleClick = (module) => {
    setSelectedModule(module);
    console.log('Selected Module:', module);
  };

  const handleReloadModuleList = async () => {
    ModuleService.getModulesByUserId(userId).then((modules) => {
      setModuleList(modules);
      console.log(modules);
    });
  };

  const handleOpenUpdateModuleWindow = () => {
    setShowModuleTable(false);
    setShowUpdateModuleWindow(true);
  };

  const [processingDelete, setProcessingDelete] = useState(false);

  const handleDeleteModule = async () => {
    try {
      setProcessingDelete(true);
      const response = await ModuleService.deleteModuleById(selectedModule.moduleId);
      if (response.status === 200) {
        handleReloadModuleList();
        toast.success('Module Deleted Successfully!');
        setSelectedModule(null);
      } else if (response.status === 400) {
        toast.error(response.data);
        toast.error('Module Deletion Failed!');
      } else {
        toast.error('Error In Module Deletion!');
      }
      setShowDeleteModuleWindow(false);
    } finally {
      setProcessingDelete(false);
    }
  };

  const handleOpenDeletePopUpWindow = () => {
    if (!selectedModule) {
      toast.error('Please select a module to delete.');
      return;
    }
    setShowDeleteModuleWindow(true);
  };

  const handleModuleCreate = async (moduleData) => {
    try {
      const response = await ModuleService.createModule(
        moduleData.moduleCode,
        moduleData.moduleName,
        moduleData.moduleEnrolmentKey,
        moduleData.semester,
        moduleData.departmentId,
        moduleData.userId
      );
      if (response.status === 200) {
        toast.success('Module Created Successfully ', response.data.moduleName);
        console.log(response);
        handleReloadModuleList();
        setShowModuleCreateWindow(false);
      } else if (response.status === 400) {
        toast.error(response.data);
      } else {
        toast.error('Something went wrong. Please try again later.');
      }
    } catch (error) {
      console.error('Fail to create modules.', error);
      toast.error('Something went wrong. Please try again later.');
    }
  };

  return (
    <div>
      <ToastContainer />
      <div className="module-Dash">
        <div className="module-SearchEvent">
          <p className="module-mainHead">Modules</p>
          <input
            type="text"
            placeholder="Search..."
            style={{
              width: '150px',
              padding: '3px 40px',
              border: '0.5px solid black',
              borderRadius: '5px',
              textAlign: 'center',
            }}
            onChange={(e) => setSearchModule(e.target.value)}
          />
        </div>
        <div className="module-ModuleList">
          <ModuleTable
            modulesList={moduleList}
            onModuleClick={handleModuleClick}
            searchModule={searchModule}
            handleDeleteModule={() => {}}
            handleOpenCreateModuleWindow={() => {}}
            handleOpenUpdateModuleWindow={() => {
              setShowUpdateModuleWindow(true);
            }}
          />

          <div className="module-List-Buttons">
            <NormalButton
              handleClick={() => {
                setShowModuleCreateWindow(true);
              }}
              title={'Create'}
              titlewithiconicon={<MdCreateNewFolder className="buttonIcon" />}
            />
            <NormalButton
              title={'Delete'}
              handleClick={handleOpenDeletePopUpWindow}
              titlewithiconicon={<RiDeleteBin5Fill className="buttonIcon" />}
            />
          </div>
        </div>
        {showUpdateModuleWindow && (
          <div
            className="Module-Create-module-Dashboard"
            handleClick={() => {
              setShowModuleCreateWindow(false);
            }}
          >
            <ModuleUpdate
              handleCloseModuleUpdateWindow={() => {
                setShowUpdateModuleWindow(false);
              }}
              handleReloadModuleList={handleReloadModuleList}
              selectedModule={selectedModule}
            />
          </div>
        )}
        {showModuleCreateWindow && (
          <div
            className="Module-Create-module-Dashboard"
            handleClick={() => {
              setShowUpdateModuleWindow(false);
            }}
          >
            <ModuleCreate
              handleCloseModuleCreateWindow={() => {
                setShowModuleCreateWindow(false);
              }}
              handleReloadModuleList={handleReloadModuleList}
              handleModuleCreate={(moduleData) => handleModuleCreate(moduleData)}
            />
          </div>
        )}
        {showDeleteModuleWindow && selectedModule && (
          <div className="module-delete-window-container">
            <WarningPopup
              handleOk={handleDeleteModule}
              handleCancel={() => {}}
              handleCloseWarningWindow={() => {
                setShowDeleteModuleWindow(false);
                handleReloadModuleList();
              }}
              titleText={'Are You Sure You Want To Delete This Module?'}
              buttonText={'Delete'}
              processing={processingDelete}
            />
          </div>
        )}
      </div>
    </div>
  );
};

export default ModulePage;
