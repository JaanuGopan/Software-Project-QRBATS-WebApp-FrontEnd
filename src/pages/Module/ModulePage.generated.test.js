import renderer from 'react-test-renderer';
import React, { useState, useEffect } from "react";
import NormalButton from "../../components/layout/AdminDashboardComponent/NormalButton";
import { MdCreateNewFolder } from "react-icons/md";
import ModuleTable from "./ModuleTable";
import ModuleService from "../../api/services/ModuleService";
import { useSelector } from "react-redux";
import { selectUser } from "../../redux/features/userSlice";
import ModuleUpdate from "./ModuleUpdate";
import { RiDeleteBin5Fill } from "react-icons/ri";

import ModuleCreate from "./ModuleCreate";
import ModulePage from "./ModulePage";

jest.mock("../../components/layout/AdminDashboardComponent/NormalButton");
jest.mock("react-icons/md");
jest.mock("./ModuleTable");
jest.mock("../../api/services/ModuleService");
jest.mock("react-redux");
jest.mock("../../redux/features/userSlice");
jest.mock("./ModuleUpdate");
jest.mock("react-icons/ri");
jest.mock("./ModulePage.css");
jest.mock("./ModuleCreate");

const renderTree = tree => renderer.create(tree);
describe('<ModulePage>', () => {
  it('should render component', () => {
    expect(renderTree(<ModulePage 
    />).toJSON()).toMatchSnapshot();
  });
  
});