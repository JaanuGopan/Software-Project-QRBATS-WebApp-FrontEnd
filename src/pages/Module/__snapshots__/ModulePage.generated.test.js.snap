// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModulePage> should render component 1`] = `
<div>
  <div
    className="module-Dash"
  >
    <div
      className="module-SearchEvent"
    >
      <p
        className="module-mainHead"
      >
        Modules
      </p>
      <input
        onChange={[Function]}
        placeholder="Search..."
        style={
          Object {
            "border": "0.5px solid black",
            "borderRadius": "5px",
            "padding": "3px 40px",
            "textAlign": "center",
            "width": "150px",
          }
        }
        type="text"
      />
    </div>
    <div
      className="module-EventList"
    >
      <div
        className="module-List-Buttons"
      />
    </div>
  </div>
</div>
`;
